modules/contextualization/cdf_file_annotation/functions/fn_file_annotation_finalize/handler.py

```diff
@@ -0,0 +1,184 @@
+    create_general_report_service,
+    create_general_retrieve_service,
+    create_general_apply_service,
+    create_general_pipeline_service,
+)
+from services.FinalizeService import AbstractFinalizeService, GeneralFinalizeService
+from services.ApplyService import IApplyService
+from services.ReportService import IReportService
+from services.RetrieveService import IRetrieveService
+from services.PipelineService import IPipelineService
+from utils.DataStructures import PerformanceTracker
+
+
+def handle(data: dict, client: CogniteClient) -> dict:
+    """
+    Main entry point for the cognite function.
+    1. Create an instance of config, logger, and tracker
+    2. Create an instance of the finalize function and create implementations of the interfaces
+    3. Run the finalize instance until...
+        4. It's been 7 minutes
+        5. There are no jobs left to process
+    6. Generate a report that includes capturing the annotations in RAW
+    NOTE: Cognite functions have a run-time limit of 10 minutes.
+    Don't want the function to die at the 10minute mark since there's no guarantee all code will execute.
+    Thus we set a timelimit of 7 minutes (conservative) so that code execution is guaranteed.
+    """
+    start_time = datetime.now(timezone.utc)
+    log_level = data.get("logLevel", "INFO")
+
+    config_instance, client = create_config_service(function_data=data, client=client)
+    logger_instance = create_logger_service(log_level)
+    tracker_instance = PerformanceTracker()
+    pipeline_instance: IPipelineService = create_general_pipeline_service(
+        client, pipeline_ext_id=data["ExtractionPipelineExtId"]
+    )
+
+    finalize_instance, report_instance = _create_finalize_service(
+        config_instance, client, logger_instance, tracker_instance
+    )
+
+    run_status: str = "success"
+    try:
+        while datetime.now(timezone.utc) - start_time < timedelta(minutes=7):
+            if finalize_instance.run():
+                return {"status": run_status, "data": data}
+            logger_instance.info(tracker_instance.generate_local_report(), "START")
+        return {"status": run_status, "data": data}
+    except Exception as e:
+        run_status = "failure"
+        msg = f"Ran into the following error: \n{str(e)}"
+        logger_instance.error(message=msg, section="BOTH")
+        return {"status": run_status, "message": msg}
+    finally:
+        result: str = report_instance.update_report()
+        logger_instance.info(result)
+        overall_report: str = tracker_instance.generate_overall_report()
+        logger_instance.info(overall_report, "BOTH")
+        # only want to report on the count of successful and failed files in ep_logs since they're relatively short
+        parts = overall_report.split("-")
+        ep_parts = parts[4:7]
+        extracted_string = " - ".join(ep_parts)
+        pipeline_instance.update_extraction_pipeline(
+            msg=f"(Finalize) {extracted_string}"

```

Comment:
So you split on a dash, then you rejoin with spaces around the dash? This is wild. Just have generate_overall_report() format the message correctly.